### ISSUE TITLE
fix: build `crane-utils` using a different toolchain

### ISSUE THREAD

**Issue by realeinherjar** [ipetkov/crane#422]
We are trying to move our CI to Nix using `crane` as the main tool, see https://github.com/bitcoindevkit/bdk/pull/1165
We have a MSRV of 1.57.0 and we need to test it on CI.
I am trying to build it using `crane.cargoBuild`:

```nix
let
  ...
  rustMSRVTarget = pkgs.rust-bin.stable."1.57.0".default;
  craneMSRVLib = (crane.mkLib pkgs).overrideToolchain rustMSRVTarget;
  MSRVArgs = {
    cargoLock = ./CargoMSRV.lock;
  };
  cargoArtifactsMSRV = craneMSRVLib.buildDepsOnly (MSRVArgs);
  ...
in
  ...
  packages = {
    MSRV = craneMSRVLib.cargoBuild (MSRVArgs // {
      cargoArtifacts = cargoArtifactsMSRV;
    });
  };
```

However I have errors:

```shell
> nix build -L .#MSRV
warning: Git tree '/Users/einherjar/git/btc/bdk' is dirty
crane-utils-deps> cargoArtifacts not set, will not reuse any cargo artifacts
crane-utils-deps> unpacking sources
crane-utils-deps> unpacking source archive /nix/store/sr1g0p7p580sz6w4f02g67c7amzkbxvn-source
crane-utils-deps> source root is source
crane-utils-deps> patching sources
crane-utils-deps> Executing configureCargoCommonVars
crane-utils-deps> updateAutotoolsGnuConfigScriptsPhase
crane-utils-deps> configuring
crane-utils-deps> will append /private/tmp/nix-build-crane-utils-deps-0.1.0.drv-0/source/.cargo-home/config.toml with contents of /nix/store/snibvbwhnwg7iqgvqlc46wx6fl6kpka2-vendor-cargo-deps/config.toml
crane-utils-deps> default configurePhase, nothing to do
crane-utils-deps> building
crane-utils-deps> ++ command cargo --version
crane-utils-deps> cargo 1.57.0 (b2e52d7ca 2021-10-21)
crane-utils-deps> ++ command cargo check --release --locked --all-targets
crane-utils-deps> error: failed to get `anyhow` as a dependency of package `crane-utils v0.1.0 (/private/tmp/nix-build-crane-utils-deps-0.1.0.drv-0/source)`
crane-utils-deps> Caused by:
crane-utils-deps>   failed to load source for dependency `anyhow`
crane-utils-deps> Caused by:
crane-utils-deps>   Unable to update registry `crates-io`
crane-utils-deps> Caused by:
crane-utils-deps>   failed to update replaced source registry `crates-io`
crane-utils-deps> Caused by:
crane-utils-deps>   failed to parse manifest at `/nix/store/snibvbwhnwg7iqgvqlc46wx6fl6kpka2-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/toml_edit-0.20.2/Cargo.toml`
crane-utils-deps> Caused by:
crane-utils-deps>   namespaced features with the `dep:` prefix are only allowed on the nightly channel and requires the `-Z namespaced-features` flag on the command-line
error: builder for '/nix/store/f3bfaxij1ja6rljsicisykl2mklqvqx2-crane-utils-deps-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       >   Unable to update registry `crates-io`
       >
       > Caused by:
       >   failed to update replaced source registry `crates-io`
       >
       > Caused by:
       >   failed to parse manifest at `/nix/store/snibvbwhnwg7iqgvqlc46wx6fl6kpka2-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/toml_edit-0.20.2/Cargo.toml`
       >
       > Caused by:
       >   namespaced features with the `dep:` prefix are only allowed on the nightly channel and requires the `-Z namespaced-features` flag on the command-line
       For full logs, run 'nix log /nix/store/f3bfaxij1ja6rljsicisykl2mklqvqx2-crane-utils-deps-0.1.0.drv'.
error: 1 dependencies of derivation '/nix/store/p3dwbi4fx6h6w3x80hhjk45b5izsjdd8-crane-utils-0.1.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/xdp693ycn3m0ygqljy7sk88m34jn8yj9-cargo-git.drv' failed to build
error: 1 dependencies of derivation '/nix/store/vhimyf39bq4a1ck90ah0lzs6ir5p8gch-linkLockedDeps.drv' failed to build
error: 1 dependencies of derivation '/nix/store/db13frhkv8rf9zcnwqpsza87512rnwzr-vendor-cargo-deps.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hnhasq7zfagd1xs3nz6iai51985ilckd-crates-build-0.1.0.drv' failed to build
```

Rust 1.57.0 cannot build `crane-utils`.

How do I override this, by making `crane-utils` be compiled by Rust latest?

**Comment by dpc** [comment]
Relevant, I guess https://github.com/ipetkov/crane/pull/402

If I implement the workaround in Flakebox, I guess it will work for Flakebox users, but I think it's a good example why it should have been handled in crane itself.

**Comment by ipetkov** [comment]
@realeinherjar using `craneMSRVLib = ((crane.mkLib pkgs).overrideToolchain rustMSRVTarget).overrideScope' (final: prev: { inherit (craneLib) craneUtils; });` should accomplish this. I tried building your branch but it failed for other reasons (couldn't find a particular crate, maybe `CargoMSRV.lock` is incorrect)

**Comment by realeinherjar** [comment]
Yes, that branch is still a mess. I need to migrate a ton of things!
However the `CargoMSRV.lock` is complete and `nix build .#MSRV` should work if we circumvent the `crane-utils` problem.

```nix
craneMSRVLib = ((crane.mkLib pkgs).overrideToolchain rustMSRVTarget).overrideScope' (final: prev: { inherit (craneLib) craneUtils; });
```

works for the `cargoArtifactsMSRV = craneMSRVLib.buildDepsOnly (commonArgs // MSRVArgs);`

However with:

```nix
craneMSRVLib.cargoBuild (commonArgs // MSRVArgs // {
  cargoArtifacts = cargoArtifactsMSRV;
  cargoLock = ./CargoMSRV.lock;
});
```

I have this error:
```shell
❱ nix build -L .#MSRV
warning: Git tree '/Users/einherjar/git/btc/bdk' is dirty
crates-build> unpacking sources
crates-build> unpacking source archive /nix/store/k97wv8swpjw3s2smh6pgsm3w0m87jxl5-source
crates-build> source root is source
crates-build> patching sources
crates-build> Executing configureCargoCommonVars
crates-build> copying cargo artifacts from /nix/store/c82xnmcr4pxjyy26q3cl123qk5l20wv7-crates-deps-0.1.0/target to target
crates-build> updateAutotoolsGnuConfigScriptsPhase
crates-build> configuring
crates-build> will append /private/tmp/nix-build-crates-build-0.1.0.drv-0/source/.cargo-home/config.toml with contents of /nix/store/l7kzzxvb4sbfz31ivj20b3hjx43075q4-vendor-cargo-deps/config.toml
crates-build> default configurePhase, nothing to do
crates-build> building
crates-build> ++ command cargo --version
crates-build> cargo 1.57.0 (b2e52d7ca 2021-10-21)
crates-build> ++ command cargo build --release --locked
crates-build> warning: /private/tmp/nix-build-crates-build-0.1.0.drv-0/source/Cargo.toml: unused manifest key: workspace.package
crates-build> error: failed to select a version for the requirement `log = "^0.4"`
crates-build> candidate versions found which didn't match: 0.4.18
crates-build> location searched: directory source `/nix/store/l7kzzxvb4sbfz31ivj20b3hjx43075q4-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0` (which is replacing registry `crates-io`)
crates-build> required by package `bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-build-0.1.0.drv-0/source/crates/bdk)`
crates-build> perhaps a crate was updated and forgotten to be re-vendored?
error: builder for '/nix/store/x11cx0sbjl99810hgggmr0m0xqd1gmhf-crates-build-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       > building
       > ++ command cargo --version
       > cargo 1.57.0 (b2e52d7ca 2021-10-21)
       > ++ command cargo build --release --locked
       > warning: /private/tmp/nix-build-crates-build-0.1.0.drv-0/source/Cargo.toml: unused manifest key: workspace.package
       > error: failed to select a version for the requirement `log = "^0.4"`
       > candidate versions found which didn't match: 0.4.18
       > location searched: directory source `/nix/store/l7kzzxvb4sbfz31ivj20b3hjx43075q4-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0` (which is replacing registry `crates-io`)
       > required by package `bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-build-0.1.0.drv-0/source/crates/bdk)`
       > perhaps a crate was updated and forgotten to be re-vendored?
       For full logs, run 'nix log /nix/store/x11cx0sbjl99810hgggmr0m0xqd1gmhf-crates-build-0.1.0.drv'.
```

It seems to ignore `CargoMSRV.lock`.
A shell from `nix develop .#MSRV` can do a `cargo check --release --locked without problems`

```shell
bash-5.2$ cargo check --release --locked
warning: /Users/einherjar/git/btc/bdk/Cargo.toml: unused manifest key: workspace.package
    Finished release [optimized] target(s) in 0.09s
```

**Comment by ipetkov** [comment]
@realeinherjar note that `cargoLock = ./CargoMSRV.lock;` only affects which crates are vendored, but you still need to instruct your actual derivation to use it and not the default `Cargo.lock` (I saw you had a commented hook that did the linking).

Also worth nothing that you might want to filter out the original `Cargo.lock` from the `MSRV` derivation so that changes to that lock file do not require (unnecessary) rebuilding

**Comment by realeinherjar** [comment]
There's the FAQ at crane.dev (https://crane.dev/faq/no-cargo-lock.html):

>  to generate your own `Cargo.lock` file and pass it in as an override by setting `cargoLock = ./path/to/Cargo.lock`. If you are calling `buildDepsOnly` or `vendorCargoDeps` directly the value must be passed there; otherwise you can pass it into `buildPackage` or `cargoBuild` and it will automatically passed through.

I think I am already doing that in the `MSRVArgs` and also I, just to be sure, put another `cargoLock = ./CargoMSRV.lock;` in the `craneMSRVLib.cargoBuild`

**Comment by ipetkov** [comment]
Ah that FAQ needs to be tweaked, thanks!

> I think I am already doing that in the MSRVArgs and also I, just to be sure, put another cargoLock = ./CargoMSRV.lock; in the craneMSRVLib.cargoBuild

This won't do anything new, you'd need some kind of `postConfigure` hook to do the actual linking. That or maybe we add our own hook that does this automatically :thinking:

**Comment by ipetkov** [comment]
Once https://github.com/ipetkov/crane/pull/425 lands any `cargoLock` value will automatically be set as the workspace `Cargo.lock` so the FAQ should be less misleading :smile:

**Comment by realeinherjar** [comment]
Thank you :)

**Comment by realeinherjar** [comment]
I am still having issues. I don't think #425 fully addresses it or if I am doing something wrong.

On a fresh shell with Rust MSRV:
```shell
❱ nix develop -L .#MSRV
nix-shell-env> cargoVendorDir not set, will not automatically configure vendored sources
bash-5.2$ rm -rf target/
bash-5.2$ mv CargoMSRV.lock Cargo.lock
bash-5.2$ cargo test --all-features
+ command cargo test --all-features
<...>
```

it runs fine.

Now if I do `nix build .#MSRV` it fails:

```nix
let
  ...
  # MSRV derivation arguments
  MSRVArgs = {
    cargoLock = ./CargoMSRV.lock;
  };
in
  ...
 packages = {
    ...
    MSRV = craneMSRVLib.cargoBuild (commonArgs // MSRVArgs // {
      cargoArtifacts = cargoArtifactsMSRV;
    });
    ...
  };
```

`cargoLock = ./CargoMSRV.lock;` is propagated through `MSRVArgs`.

**Comment by ipetkov** [comment]
@realeinherjar tell you what, try to get the build working (inside of the sandbox) without the `Cargo.lock` juggling (meaning change the source so `Cargo.lock` has the MSRV versions).

Once that is good, copy the `Cargo.lock` to `CargoMSRV.lock` and try again, and if it still doesn't work feel free to reopen the issue. Then we can check what's different between those two setups way more easily

**Comment by realeinherjar** [comment]
Ok something funky is going on. Moving `CargoMSRV.lock` to `Cargo.lock` and
```nix
        MSRVArgs = {
          cargoLock = ./Cargo.lock;
        };
```
and doing `nix build .#MSRV` runs clean (well, actually, not clean because there is a test failure that is the same in Rust latest, but it compiles which didn't before).

So something is still going on that #425 did not address...

**Comment by ipetkov** [comment]
If you do `MSRVArgs = {};` with `CargoMSRV.lock` moved to `Cargo.lock` does it still build fine?

**Comment by realeinherjar** [comment]
Ok I did `MSRVArgs = {};` with `CargoMSRV.lock` moved to `Cargo.lock`.
Still builds!
Here's the full log:

```shell
 I  ~/g/b/bdk einherjar/nix• 14m 51.2s | 101 ❱ mv CargoMSRV.lock Cargo.lock
 I  ~/g/b/bdk einherjar/nix• ❱ nix build -L --keep-failed .#MSRV
warning: Git tree '/Users/einherjar/git/btc/bdk' is dirty
crates-deps> cargoArtifacts not set, will not reuse any cargo artifacts
crates-deps> unpacking sources
crates-deps> unpacking source archive /nix/store/kqpvvhg7jvxc0z6657cgjkri9b03hjxs-source
crates-deps> source root is source
crates-deps> patching sources
crates-deps> Executing configureCargoCommonVars
crates-deps> updateAutotoolsGnuConfigScriptsPhase
crates-deps> configuring
crates-deps> will append /private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/.cargo-home/config.toml with contents of /nix/store/4xqj2z2xhckggpvjsfxlca2cjnghjm7y-vendor-cargo-deps/config.toml
crates-deps> default configurePhase, nothing to do
crates-deps> building
crates-deps> ++ command cargo --version
crates-deps> cargo 1.57.0 (b2e52d7ca 2021-10-21)
crates-deps> ++ command cargo check --release --locked --all-targets
crates-deps> warning: /private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/Cargo.toml: unused manifest key: workspace.package
crates-deps>    Compiling libc v0.2.149
crates-deps>    Compiling proc-macro2 v1.0.69
crates-deps>    Compiling unicode-ident v1.0.12
crates-deps>     Checking cfg-if v1.0.0
crates-deps>    Compiling serde v1.0.189
crates-deps>    Compiling bitcoin-private v0.1.0
crates-deps>     Checking ppv-lite86 v0.2.17
crates-deps>     Checking base64 v0.13.1
crates-deps>    Compiling hex_lit v0.1.1
crates-deps>    Compiling bitcoin v0.30.1
crates-deps>     Checking bech32 v0.9.1
crates-deps>    Compiling autocfg v1.1.0
crates-deps>     Checking itoa v1.0.9
crates-deps>    Compiling bdk_chain v0.6.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/chain)
crates-deps>     Checking ryu v1.0.15
crates-deps>    Compiling serde_json v1.0.107
crates-deps>    Compiling log v0.4.18
crates-deps>     Checking bitflags v1.3.2
crates-deps>     Checking untrusted v0.7.1
crates-deps>     Checking once_cell v1.17.2
crates-deps>     Checking hashbrown v0.12.3
crates-deps>    Compiling memchr v2.5.0
crates-deps>    Compiling bdk_file_store v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/file_store)
crates-deps>     Checking byteorder v1.4.3
crates-deps>    Compiling io-lifetimes v1.0.11
crates-deps>    Compiling rustix v0.37.25
crates-deps>     Checking pin-project-lite v0.2.13
crates-deps>    Compiling futures-core v0.3.28
crates-deps>    Compiling anyhow v1.0.75
crates-deps>     Checking fastrand v1.9.0
crates-deps>     Checking pin-utils v0.1.0
crates-deps>     Checking bytes v1.5.0
crates-deps>     Checking futures-sink v0.3.28
crates-deps>    Compiling futures-task v0.3.28
crates-deps>    Compiling futures-channel v0.3.28
crates-deps>    Compiling futures-util v0.3.28
crates-deps>     Checking futures-io v0.3.28
crates-deps>     Checking lazy_static v1.4.0
crates-deps>    Compiling maybe-uninit v2.0.0
crates-deps>    Compiling version_check v0.9.4
crates-deps>     Checking either v1.9.0
crates-deps>     Checking core-foundation-sys v0.8.4
crates-deps>     Checking fnv v1.0.7
crates-deps>     Checking termcolor v1.3.0
crates-deps>     Checking percent-encoding v2.3.0
crates-deps>    Compiling httparse v1.8.0
crates-deps>    Compiling native-tls v0.2.11
crates-deps>    Compiling syn v1.0.109
crates-deps>     Checking matches v0.1.10
crates-deps>     Checking try-lock v0.2.4
crates-deps>    Compiling crc32fast v1.3.2
crates-deps>    Compiling thiserror v1.0.49
crates-deps>    Compiling bdk_tmp_plan v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/tmp_plan)
crates-deps>     Checking unicode-bidi v0.3.13
crates-deps>    Compiling bdk_coin_select v0.0.1 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/coin_select)
crates-deps>    Compiling rustls v0.20.8
crates-deps>     Checking adler v1.0.2
crates-deps>     Checking httpdate v1.0.3
crates-deps>     Checking tower-service v0.3.2
crates-deps>    Compiling bitcoin-internals v0.1.0
crates-deps>    Compiling bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bdk)
crates-deps>    Compiling async-trait v0.1.74
crates-deps>     Checking ipnet v2.8.0
crates-deps>    Compiling heck v0.4.1
crates-deps>     Checking os_str_bytes v6.5.1
crates-deps>     Checking mime v0.3.17
crates-deps>     Checking base64 v0.21.4
crates-deps>     Checking textwrap v0.16.0
crates-deps>    Compiling bdk_esplora v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/esplora)
crates-deps>     Checking strsim v0.10.0
crates-deps>    Compiling example_cli v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_cli)
crates-deps>    Compiling rustls v0.21.1
crates-deps>    Compiling bdk_electrum v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/electrum)
crates-deps>    Compiling bdk_bitcoind_rpc v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bitcoind_rpc)
crates-deps>    Compiling winapi-build v0.1.1
crates-deps>     Checking winapi v0.2.8
crates-deps>     Checking quick-error v1.2.3
crates-deps>     Checking regex-syntax v0.6.29
crates-deps>    Compiling example_esplora v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_esplora)
crates-deps>    Compiling wallet_esplora_async v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_async)
crates-deps>    Compiling example_bitcoind_rpc_polling v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_bitcoind_rpc_polling)
crates-deps>    Compiling example_electrum v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_electrum)
crates-deps>    Compiling wallet_esplora_blocking v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_blocking)
crates-deps>    Compiling wallet_electrum_example v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_electrum)
crates-deps>    Compiling electrsd v0.25.0 (https://github.com/realeinherjar/electrsd?branch=einherjar/optional-download#d3984868)
crates-deps>     Checking assert_matches v1.5.0
crates-deps>     Checking encoding_rs v0.8.33
crates-deps>    Compiling indexmap v1.9.3
crates-deps>    Compiling tempfile v3.6.0
crates-deps>    Compiling slab v0.4.9
crates-deps>    Compiling tokio v1.29.1
crates-deps>    Compiling memoffset v0.6.5
crates-deps>     Checking tracing-core v0.1.32
crates-deps>     Checking http v0.2.9
crates-deps>     Checking form_urlencoded v1.2.0
crates-deps>     Checking want v0.3.1
crates-deps>    Compiling proc-macro-error-attr v1.0.4
crates-deps>    Compiling proc-macro-error v1.0.4
crates-deps>     Checking miniz_oxide v0.7.1
crates-deps>     Checking clap_lex v0.2.4
crates-deps>     Checking humantime v1.3.0
crates-deps>    Compiling kernel32-sys v0.2.2
crates-deps>     Checking tracing v0.1.39
crates-deps>     Checking http-body v0.4.5
crates-deps>    Compiling quote v1.0.33
crates-deps>    Compiling cc v1.0.83
crates-deps>     Checking aho-corasick v0.7.20
crates-deps>     Checking getrandom v0.2.10
crates-deps>     Checking errno v0.3.5
crates-deps>     Checking num_cpus v1.16.0
crates-deps>     Checking mio v0.8.8
crates-deps>     Checking socket2 v0.4.9
crates-deps>     Checking core-foundation v0.9.3
crates-deps>     Checking atty v0.2.14
crates-deps>     Checking security-framework-sys v2.9.1
crates-deps>     Checking socks v0.3.4
crates-deps>     Checking which v4.4.0
crates-deps>    Compiling bitcoind v0.33.2
crates-deps>     Checking smallvec v0.6.14
crates-deps>     Checking flate2 v1.0.26
crates-deps>    Compiling syn v2.0.38
crates-deps>     Checking rand_core v0.6.4
crates-deps>     Checking regex v1.7.3
crates-deps>     Checking security-framework v2.8.1
crates-deps>    Compiling secp256k1-sys v0.8.1
crates-deps>    Compiling ring v0.16.20
crates-deps>     Checking unicode-normalization v0.1.9
crates-deps>     Checking rand_chacha v0.3.1
crates-deps>     Checking nix v0.25.1
crates-deps>     Checking env_logger v0.7.1
crates-deps>     Checking ctrlc v2.0.1
crates-deps>     Checking rand v0.8.5
crates-deps>     Checking idna v0.2.1
crates-deps>     Checking url v2.3.0
crates-deps>    Compiling clap_derive v3.2.25
crates-deps>    Compiling serde_derive v1.0.189
crates-deps>    Compiling tokio-macros v2.1.0
crates-deps>    Compiling futures-macro v0.3.28
crates-deps>    Compiling thiserror-impl v1.0.49
crates-deps>     Checking clap v3.2.25
crates-deps>     Checking tokio-util v0.7.8
crates-deps>     Checking tokio-native-tls v0.3.1
crates-deps>     Checking tokio-socks v0.5.1
crates-deps>     Checking futures-executor v0.3.28
crates-deps>     Checking futures v0.3.28
crates-deps>     Checking h2 v0.3.20
crates-deps>     Checking bitcoin_hashes v0.12.0
crates-deps>     Checking bincode v1.3.3
crates-deps>     Checking serde_urlencoded v0.7.1
crates-deps>     Checking hyper v0.14.27
crates-deps>     Checking jsonrpc v0.14.1
crates-deps>     Checking secp256k1 v0.27.0
crates-deps>     Checking webpki v0.22.2
crates-deps>     Checking sct v0.7.0
crates-deps>     Checking rustls-webpki v0.100.1
crates-deps>     Checking webpki-roots v0.22.6
crates-deps>     Checking hyper-tls v0.5.0
crates-deps>     Checking reqwest v0.11.18
crates-deps>     Checking ureq v2.6.2
crates-deps>     Checking miniscript v10.0.0
crates-deps>     Checking esplora-client v0.6.0
crates-deps>     Checking bitcoincore-rpc-json v0.17.0
crates-deps>     Checking electrum-client v0.18.0
crates-deps>     Checking electrum-client v0.15.1
crates-deps>     Checking bitcoincore-rpc v0.17.0
crates-deps>     Finished release [optimized] target(s) in 27.82s
crates-deps> ++ command cargo build --release --locked
crates-deps> warning: /private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/Cargo.toml: unused manifest key: workspace.package
crates-deps>    Compiling cfg-if v1.0.0
crates-deps>    Compiling ppv-lite86 v0.2.17
crates-deps>    Compiling base64 v0.13.1
crates-deps>    Compiling itoa v1.0.9
crates-deps>    Compiling bech32 v0.9.1
crates-deps>    Compiling ryu v1.0.15
crates-deps>    Compiling log v0.4.18
crates-deps>    Compiling untrusted v0.7.1
crates-deps>    Compiling once_cell v1.17.2
crates-deps>    Compiling bitflags v1.3.2
crates-deps>    Compiling pin-project-lite v0.2.13
crates-deps>    Compiling hashbrown v0.12.3
crates-deps>    Compiling futures-sink v0.3.28
crates-deps>    Compiling bytes v1.5.0
crates-deps>    Compiling byteorder v1.4.3
crates-deps>    Compiling futures-io v0.3.28
crates-deps>    Compiling pin-utils v0.1.0
crates-deps>    Compiling core-foundation-sys v0.8.4
crates-deps>    Compiling fnv v1.0.7
crates-deps>    Compiling fastrand v1.9.0
crates-deps>    Compiling percent-encoding v2.3.0
crates-deps>    Compiling try-lock v0.2.4
crates-deps>    Compiling lazy_static v1.4.0
crates-deps>    Compiling unicode-bidi v0.3.13
crates-deps>    Compiling matches v0.1.10
crates-deps>    Compiling httpdate v1.0.3
crates-deps>    Compiling tower-service v0.3.2
crates-deps>    Compiling adler v1.0.2
crates-deps>    Compiling either v1.9.0
crates-deps>    Compiling base64 v0.21.4
crates-deps>    Compiling ipnet v2.8.0
crates-deps>    Compiling mime v0.3.17
crates-deps>    Compiling os_str_bytes v6.5.1
crates-deps>    Compiling strsim v0.10.0
crates-deps>    Compiling termcolor v1.3.0
crates-deps>    Compiling textwrap v0.16.0
crates-deps>    Compiling winapi v0.2.8
crates-deps>    Compiling encoding_rs v0.8.33
crates-deps>    Compiling tracing-core v0.1.32
crates-deps>    Compiling form_urlencoded v1.2.0
crates-deps>    Compiling http v0.2.9
crates-deps>    Compiling want v0.3.1
crates-deps>    Compiling miniz_oxide v0.7.1
crates-deps>    Compiling clap_lex v0.2.4
crates-deps>    Compiling libc v0.2.149
crates-deps>    Compiling bitcoin-private v0.1.0
crates-deps>    Compiling hex_lit v0.1.1
crates-deps>    Compiling futures-core v0.3.28
crates-deps>    Compiling futures-task v0.3.28
crates-deps>    Compiling memchr v2.5.0
crates-deps>    Compiling maybe-uninit v2.0.0
crates-deps>    Compiling httparse v1.8.0
crates-deps>    Compiling tracing v0.1.39
crates-deps>    Compiling crc32fast v1.3.2
crates-deps>    Compiling bitcoin-internals v0.1.0
crates-deps>    Compiling anyhow v1.0.75
crates-deps>    Compiling http-body v0.4.5
crates-deps>    Compiling indexmap v1.9.3
crates-deps>    Compiling slab v0.4.9
crates-deps>    Compiling futures-channel v0.3.28
crates-deps>    Compiling smallvec v0.6.14
crates-deps>    Compiling flate2 v1.0.26
crates-deps>    Compiling getrandom v0.2.10
crates-deps>    Compiling num_cpus v1.16.0
crates-deps>    Compiling socket2 v0.4.9
crates-deps>    Compiling mio v0.8.8
crates-deps>    Compiling io-lifetimes v1.0.11
crates-deps>    Compiling errno v0.3.5
crates-deps>    Compiling security-framework-sys v2.9.1
crates-deps>    Compiling core-foundation v0.9.3
crates-deps>    Compiling socks v0.3.4
crates-deps>    Compiling atty v0.2.14
crates-deps>    Compiling kernel32-sys v0.2.2
crates-deps>    Compiling unicode-normalization v0.1.9
crates-deps>    Compiling rand_core v0.6.4
crates-deps>    Compiling rustix v0.37.25
crates-deps>    Compiling security-framework v2.8.1
crates-deps>    Compiling ctrlc v2.0.1
crates-deps>    Compiling rand_chacha v0.3.1
crates-deps>    Compiling serde v1.0.189
crates-deps>    Compiling futures-util v0.3.28
crates-deps>    Compiling tokio v1.29.1
crates-deps>    Compiling thiserror v1.0.49
crates-deps>    Compiling secp256k1-sys v0.8.1
crates-deps>    Compiling ring v0.16.20
crates-deps>    Compiling idna v0.2.1
crates-deps>    Compiling rand v0.8.5
crates-deps>    Compiling tempfile v3.6.0
crates-deps>    Compiling clap v3.2.25
crates-deps>    Compiling url v2.3.0
crates-deps>    Compiling sct v0.7.0
crates-deps>    Compiling webpki v0.22.2
crates-deps>    Compiling rustls-webpki v0.100.1
crates-deps>    Compiling futures-executor v0.3.28
crates-deps>    Compiling bitcoin_hashes v0.12.0
crates-deps>    Compiling serde_json v1.0.107
crates-deps>    Compiling bincode v1.3.3
crates-deps>    Compiling serde_urlencoded v0.7.1
crates-deps>    Compiling native-tls v0.2.11
crates-deps>    Compiling tokio-util v0.7.8
crates-deps>    Compiling tokio-socks v0.5.1
crates-deps>    Compiling webpki-roots v0.22.6
crates-deps>    Compiling rustls v0.20.8
crates-deps>    Compiling futures v0.3.28
crates-deps>    Compiling rustls v0.21.1
crates-deps>    Compiling secp256k1 v0.27.0
crates-deps>    Compiling jsonrpc v0.14.1
crates-deps>    Compiling tokio-native-tls v0.3.1
crates-deps>    Compiling h2 v0.3.20
crates-deps>    Compiling bitcoin v0.30.1
crates-deps>    Compiling ureq v2.6.2
crates-deps>    Compiling hyper v0.14.27
crates-deps>    Compiling miniscript v10.0.0
crates-deps>    Compiling electrum-client v0.18.0
crates-deps>    Compiling bitcoincore-rpc-json v0.17.0
crates-deps>    Compiling hyper-tls v0.5.0
crates-deps>    Compiling bdk_chain v0.6.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/chain)
crates-deps>    Compiling reqwest v0.11.18
crates-deps>    Compiling bdk_file_store v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/file_store)
crates-deps>    Compiling bdk_tmp_plan v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/tmp_plan)
crates-deps>    Compiling bdk_coin_select v0.0.1 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/coin_select)
crates-deps>    Compiling bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bdk)
crates-deps>    Compiling bdk_electrum v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/electrum)
crates-deps>    Compiling example_cli v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_cli)
crates-deps>    Compiling wallet_electrum_example v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_electrum)
crates-deps>    Compiling esplora-client v0.6.0
crates-deps>    Compiling example_electrum v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_electrum)
crates-deps>    Compiling bdk_esplora v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/esplora)
crates-deps>    Compiling bitcoincore-rpc v0.17.0
crates-deps>    Compiling example_esplora v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_esplora)
crates-deps>    Compiling wallet_esplora_blocking v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_blocking)
crates-deps>    Compiling wallet_esplora_async v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_async)
crates-deps>    Compiling bdk_bitcoind_rpc v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bitcoind_rpc)
crates-deps>    Compiling example_bitcoind_rpc_polling v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_bitcoind_rpc_polling)
crates-deps>     Finished release [optimized] target(s) in 39.46s
crates-deps> buildPhase completed in 1 minutes 7 seconds
crates-deps> running tests
crates-deps> ++ command cargo test --release --locked
crates-deps> warning: /private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/Cargo.toml: unused manifest key: workspace.package
crates-deps>    Compiling quick-error v1.2.3
crates-deps>    Compiling regex-syntax v0.6.29
crates-deps>    Compiling assert_matches v1.5.0
crates-deps>    Compiling log v0.4.18
crates-deps>    Compiling which v4.4.0
crates-deps>    Compiling aho-corasick v0.7.20
crates-deps>    Compiling memoffset v0.6.5
crates-deps>    Compiling bdk_chain v0.6.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/chain)
crates-deps>    Compiling bdk_coin_select v0.0.1 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/coin_select)
crates-deps>    Compiling bdk_file_store v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/file_store)
crates-deps>    Compiling bdk_tmp_plan v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/nursery/tmp_plan)
crates-deps>    Compiling nix v0.25.1
crates-deps>    Compiling example_cli v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_cli)
crates-deps>    Compiling humantime v1.3.0
crates-deps>    Compiling rustls v0.20.8
crates-deps>    Compiling reqwest v0.11.18
crates-deps>    Compiling bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bdk)
crates-deps>    Compiling bitcoincore-rpc v0.17.0
crates-deps>    Compiling rustls v0.21.1
crates-deps>    Compiling electrum-client v0.15.1
crates-deps>    Compiling bitcoind v0.33.2
crates-deps>    Compiling bdk_bitcoind_rpc v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/bitcoind_rpc)
crates-deps>    Compiling regex v1.7.3
crates-deps>    Compiling ureq v2.6.2
crates-deps>    Compiling example_bitcoind_rpc_polling v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_bitcoind_rpc_polling)
crates-deps>    Compiling electrsd v0.25.0 (https://github.com/realeinherjar/electrsd?branch=einherjar/optional-download#d3984868)
crates-deps>    Compiling electrum-client v0.18.0
crates-deps>    Compiling env_logger v0.7.1
crates-deps>    Compiling esplora-client v0.6.0
crates-deps>    Compiling bdk_electrum v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/electrum)
crates-deps>    Compiling bdk_esplora v0.4.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/crates/esplora)
crates-deps>    Compiling wallet_electrum_example v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_electrum)
crates-deps>    Compiling example_electrum v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_electrum)
crates-deps>    Compiling wallet_esplora_blocking v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_blocking)
crates-deps>    Compiling wallet_esplora_async v0.2.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/wallet_esplora_async)
crates-deps>    Compiling example_esplora v0.1.0 (/private/tmp/nix-build-crates-deps-0.1.0.drv-2/source/example-crates/example_esplora)
crates-deps>     Finished release [optimized] target(s) in 17.64s
crates-deps>      Running unittests (target/release/deps/bdk-926a2000cb94ee6a)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk-590a03c592317323)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_bitcoind_rpc-a941495b7ab06914)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_bitcoind_rpc-3696e8f2d81dcd89)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_chain-9e827a118ac0c33d)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_chain-cc1809d208926c8b)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_coin_select-8febc199bb0eeeef)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_coin_select-d6c850cbec502a51)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_electrum-09700b7e1db9e758)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_electrum-690650a2146d76c7)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_esplora-867e01eb32ba12d1)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_esplora-cdd9f01aeb49c660)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_file_store-9d0713c67210f668)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_file_store-95ced1c395a31b32)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/bdk_tmp_plan-c78d354d539e5815)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_bdk_tmp_plan-90bbd3386031533b)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/example_bitcoind_rpc_polling-beac086308ffbc4b)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_example_bitcoind_rpc_polling-a28f0ef0a9ab3a0c)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/example_cli-b50b3276169272cd)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_example_cli-2b7af038a2b49b48)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/example_electrum-1fde1be481e537ee)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_example_electrum-c3c35baaaa43ae44)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/example_esplora-4ed494947c07561b)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_example_esplora-c2ef7c0861934423)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/wallet_electrum_example-7fae224e5bd66504)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_wallet_electrum_example-cda48fab0bf74bce)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/wallet_esplora_async-2c7a5e2cf961de24)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_wallet_esplora_async-580fa8d159c36e92)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/wallet_esplora_blocking-8bb0ae90ca1d136b)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>      Running unittests (target/release/deps/crane_dummy_wallet_esplora_blocking-9e3d2b8459ec6a80)
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_bitcoind_rpc
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_chain
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_coin_select
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_electrum
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_esplora
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_file_store
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests bdk_tmp_plan
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests example_bitcoind_rpc_polling
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests example_cli
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests example_electrum
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests example_esplora
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests wallet_electrum_example
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests wallet_esplora_async
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps>    Doc-tests wallet_esplora_blocking
crates-deps> running 0 tests
crates-deps> test result: ok(B. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
crates-deps> installing
crates-deps> installing target to /nix/store/zfp19klxdz4ld29kzn05gs9xa420rihm-crates-deps-0.1.0
crates-deps> post-installation fixup
crates-deps> checking for references to /private/tmp/nix-build-crates-deps-0.1.0.drv-2/ in /nix/store/zfp19klxdz4ld29kzn05gs9xa420rihm-crates-deps-0.1.0...
crates-deps> patching script interpreter paths in /nix/store/zfp19klxdz4ld29kzn05gs9xa420rihm-crates-deps-0.1.0
crates-build> unpacking sources
crates-build> unpacking source archive /nix/store/wnqyhsiwl0h5imd6cpsms3rrwr9lvhv8-source
crates-build> source root is source
crates-build> patching sources
crates-build> Executing configureCargoCommonVars
crates-build> copying cargo artifacts from /nix/store/zfp19klxdz4ld29kzn05gs9xa420rihm-crates-deps-0.1.0/target to target
crates-build> updateAutotoolsGnuConfigScriptsPhase
crates-build> configuring
crates-build> will append /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/.cargo-home/config.toml with contents of /nix/store/4xqj2z2xhckggpvjsfxlca2cjnghjm7y-vendor-cargo-deps/config.toml
crates-build> default configurePhase, nothing to do
crates-build> building
crates-build> ++ command cargo --version
crates-build> cargo 1.57.0 (b2e52d7ca 2021-10-21)
crates-build> ++ command cargo build --release --locked
crates-build> warning: /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/Cargo.toml: unused manifest key: workspace.package
crates-build>    Compiling bdk_chain v0.6.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/chain)
crates-build>    Compiling bdk_file_store v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/file_store)
crates-build>    Compiling bdk_tmp_plan v0.1.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/nursery/tmp_plan)
crates-build>    Compiling bdk_coin_select v0.0.1 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/nursery/coin_select)
crates-build>    Compiling bdk v1.0.0-alpha.2 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/bdk)
crates-build>    Compiling bdk_esplora v0.4.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/esplora)
crates-build>    Compiling bdk_electrum v0.4.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/electrum)
crates-build>    Compiling bdk_bitcoind_rpc v0.1.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/crates/bitcoind_rpc)
crates-build>    Compiling example_cli v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/example_cli)
crates-build>    Compiling example_esplora v0.1.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/example_esplora)
crates-build>    Compiling example_bitcoind_rpc_polling v0.1.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/example_bitcoind_rpc_polling)
crates-build>    Compiling example_electrum v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/example_electrum)
crates-build>    Compiling wallet_esplora_async v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/wallet_esplora_async)
crates-build>    Compiling wallet_electrum_example v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/wallet_electrum)
crates-build>    Compiling wallet_esplora_blocking v0.2.0 (/private/tmp/nix-build-crates-build-0.1.0.drv-8/source/example-crates/wallet_esplora_blocking)
crates-build>     Finished release [optimized] target(s) in 45.91s
crates-build> buildPhase completed in 46 seconds
crates-build> installing
crates-build> symlinking duplicates in target to /nix/store/zfp19klxdz4ld29kzn05gs9xa420rihm-crates-deps-0.1.0/target
crates-build> installing target to /nix/store/rdw9vdzjb7fh5fnp9m9fx91zc3zb6mic-crates-build-0.1.0
crates-build> post-installation fixup
crates-build> checking for references to /private/tmp/nix-build-crates-build-0.1.0.drv-8/ in /nix/store/rdw9vdzjb7fh5fnp9m9fx91zc3zb6mic-crates-build-0.1.0...
crates-build> patching script interpreter paths in /nix/store/rdw9vdzjb7fh5fnp9m9fx91zc3zb6mic-crates-build-0.1.0
 N  ~/g/b/bdk einherjar/nix• 2m 54.3s ❱
```

**Comment by ipetkov** [comment]
Okay that's good! What are the logs when trying to build with `MSRV = { cargoLock = ./CargoMSRV.lock; }`? (might want to _delete_ `Cargo.lock` when testing this, if it builds fine but doesn't when `Cargo.lock` is brought back we have another hint)

**Comment by realeinherjar** [comment]
Ha! When we remove `Cargo.lock` it works with `CargoMSRV.lock`.

I am not getting any of the warnings in `lib/setupHooks/replaceCargoLockHook.sh`, e.g.
https://github.com/ipetkov/crane/blob/a26f215d943c1649024aa40860c1c4958a82b47d/lib/setupHooks/replaceCargoLockHook.sh#L5

**Comment by ipetkov** [comment]
Interesting, thanks for confirming! I wonder if something ends up trying to use `./Cargo.lock` instead of honoring the override :thinking: 

Here's a random thing to try: break `Cargo.lock` (e.g. paste some malformed text in there that is obviously invalid TOML) and let's see what breaks, that might be a narrower clue!

**Comment by dpc** [comment]
> @realeinherjar using `craneMSRVLib = ((crane.mkLib pkgs).overrideToolchain rustMSRVTarget).overrideScope' (final: prev: { inherit (craneLib) craneUtils; });` should accomplish this. I tried building your branch but it failed for other reasons (couldn't find a particular crate, maybe `CargoMSRV.lock` is incorrect)


While this works, it's a hard to come up with (and understand) workaround (hack?).

IMO, building purely internal binary with the same toolchain that the user is supposed to configure and use is just bad architecture and asking for trouble.

`craneUtils` should be built "by the outside". Possibly even using rustc and `buildRustPackage` from the `nixpkgs` to avoid pulling in yet another toolchain.

@ipetkov

**Comment by realeinherjar** [comment]
> Here's a random thing to try: break `Cargo.lock` (e.g. paste some malformed text in there that is obviously invalid TOML) and let's see what breaks, that might be a narrower clue!

@ipetkov here's the error:
```shell
crates-build> default configurePhase, nothing to do
crates-build> building
crates-build> ++ command cargo --version
crates-build> cargo 1.57.0 (b2e52d7ca 2021-10-21)
crates-build> ++ command cargo build --release --locked
crates-build> warning: /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/Cargo.toml: unused manifest key: workspace.package
crates-build> error: failed to parse lock file at: /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/Cargo.lock
crates-build> Caused by:
crates-build>   could not parse input as TOML
crates-build> Caused by:
crates-build>   expected an equals, found a newline at line 3 column 23
error: builder for '/nix/store/v6dhzpkkbl8gc9n2qky7viipzixj8jdb-crates-build-0.1.0.drv' failed with exit code 101;
       last 10 log lines:
       > cargo 1.57.0 (b2e52d7ca 2021-10-21)
       > ++ command cargo build --release --locked
       > warning: /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/Cargo.toml: unused manifest key: workspace.package
       > error: failed to parse lock file at: /private/tmp/nix-build-crates-build-0.1.0.drv-8/source/Cargo.lock
       >
       > Caused by:
       >   could not parse input as TOML
       >
       > Caused by:
       >   expected an equals, found a newline at line 3 column 23
       For full logs, run 'nix log /nix/store/v6dhzpkkbl8gc9n2qky7viipzixj8jdb-crates-build-0.1.0.drv'.
```

**Comment by dpc** [comment]
@realeinherjar So the old (now broken by you) `Cargo.lock` is still being used, even with the crane change that was supposed to overwrite it? Do I get it right?

A thing to double check is if you don't have any existing build that would still use it and fail before you get to the point where replacement is requested.

**Comment by realeinherjar** [comment]
> @realeinherjar So the old (now broken by you) `Cargo.lock` is still being used, even with the crane change that was supposed to overwrite it? Do I get it right?

yes

**Comment by ipetkov** [comment]
@realeinherjar hmm weird that the cargoLock overwrite hook isn't running :thinking: 

---

@dpc 

> IMO, building purely internal binary with the same toolchain that is meant for the user to configure and use is just bad architecture and asking for trouble, and for not good reason really. Users shouldn't need workarounds just because some internal tooling can't be built with configuration they need to use.
> 
> `craneUtils` should be built "by the outside". Possibly even using rustc and `buildRustPackage` from the `nixpkgs` to avoid pulling in yet another toolchain (technically it will be another rustc anyway, but at least standard one coming from `nixpkgs`, possibly shared with existing stuff).

Sorry for deciding to dogfood there :man_shrugging: 

Should be fixed in https://github.com/ipetkov/crane/pull/431

**Comment by ipetkov** [comment]
Figured out the issue: copying the overridden `Cargo.lock` cannot be done as a `postPatch` hook because the builder does not `cd`  into the unpacked source until after those hooks are done (thus we effectively end up linking the `Cargo.lock` to `..` when the build runs).

The fix is to change the hook to `preConfigure` and then it seems to build fine. Will put a PR up in a minute

**Comment by ipetkov** [comment]
Should be fixed with https://github.com/ipetkov/crane/pull/434!

**Comment by realeinherjar** [comment]
Now it works! Thank you!
